File: javaswing-lab/src/main/java/org/example/model/RandomShapeFactory.java
package org.example.model;

import org.example.common.GISTestConfig;

import java.awt.Color;
import java.util.Random;

public final class RandomShapeFactory {
    
    private RandomShapeFactory() {
    }
    
    public static GISPoint randomPoint(Random random) {
        double x = randomCoord(random, GISTestConfig.WORLD_MIN_X, GISTestConfig.WORLD_MAX_X);
        double y = randomCoord(random, GISTestConfig.WORLD_MIN_Y, GISTestConfig.WORLD_MAX_Y);
        Color color = randomColor(random, 120, 220);
        int size = 3 + random.nextInt(5);
        return new GISPoint(x, y, color, size);
    }
    
    public static GISLine randomLine(Random random) {
        double x1 = randomCoord(random, GISTestConfig.WORLD_MIN_X, GISTestConfig.WORLD_MAX_X);
        double y1 = randomCoord(random, GISTestConfig.WORLD_MIN_Y, GISTestConfig.WORLD_MAX_Y);
        double x2 = randomCoord(random, GISTestConfig.WORLD_MIN_X, GISTestConfig.WORLD_MAX_X);
        double y2 = randomCoord(random, GISTestConfig.WORLD_MIN_Y, GISTestConfig.WORLD_MAX_Y);
        Color color = randomColor(random, 80, 200);
        float strokeWidth = 1.0f + random.nextFloat() * 2.0f;
        return new GISLine(x1, y1, x2, y2, color, strokeWidth);
    }
    
    public static GISPolygon randomPolygon(Random random) {
        int points = 3 + random.nextInt(4);
        double[] xPoints = new double[points];
        double[] yPoints = new double[points];
        double centerX = randomCoord(random, GISTestConfig.WORLD_MIN_X, GISTestConfig.WORLD_MAX_X);
        double centerY = randomCoord(random, GISTestConfig.WORLD_MIN_Y, GISTestConfig.WORLD_MAX_Y);
        double radius = 20 + random.nextDouble() * 80;
        
        for (int i = 0; i < points; i++) {
            double angle = 2 * Math.PI * i / points + random.nextDouble() * 0.3;
            xPoints[i] = centerX + Math.cos(angle) * radius;
            yPoints[i] = centerY + Math.sin(angle) * radius;
        }
        
        Color fillColor = randomColor(random, 80, 180);
        Color strokeColor = randomColor(random, 60, 160);
        float strokeWidth = 1.0f + random.nextFloat() * 2.5f;
        return new GISPolygon(xPoints, yPoints, fillColor, strokeColor, strokeWidth);
    }
    
    private static double randomCoord(Random random, double min, double max) {
        return min + (max - min) * random.nextDouble();
    }
    
    private static Color randomColor(Random random, int min, int max) {
        int r = min + random.nextInt(max - min + 1);
        int g = min + random.nextInt(max - min + 1);
        int b = min + random.nextInt(max - min + 1);
        return new Color(r, g, b);
    }
}

File: javafx-lab/src/main/kotlin/org/example/model/RandomShapeFactory.kt
package org.example.model

import javafx.scene.paint.Color
import org.example.common.GISTestConfig
import kotlin.random.Random

object RandomShapeFactory {
    fun randomPoint(random: Random): GISPoint {
        val x = randomCoord(random, GISTestConfig.WORLD_MIN_X, GISTestConfig.WORLD_MAX_X)
        val y = randomCoord(random, GISTestConfig.WORLD_MIN_Y, GISTestConfig.WORLD_MAX_Y)
        val color = randomColor(random, 120, 220, 1.0)
        val size = 3 + random.nextInt(5)
        return GISPoint(x, y, color, size)
    }

    fun randomLine(random: Random): GISLine {
        val x1 = randomCoord(random, GISTestConfig.WORLD_MIN_X, GISTestConfig.WORLD_MAX_X)
        val y1 = randomCoord(random, GISTestConfig.WORLD_MIN_Y, GISTestConfig.WORLD_MAX_Y)
        val x2 = randomCoord(random, GISTestConfig.WORLD_MIN_X, GISTestConfig.WORLD_MAX_X)
        val y2 = randomCoord(random, GISTestConfig.WORLD_MIN_Y, GISTestConfig.WORLD_MAX_Y)
        val color = randomColor(random, 80, 200, 1.0)
        val strokeWidth = 1.0 + random.nextDouble() * 2.0
        return GISLine(x1, y1, x2, y2, color, strokeWidth)
    }

    fun randomPolygon(random: Random): GISPolygon {
        val points = 3 + random.nextInt(4)
        val xPoints = DoubleArray(points)
        val yPoints = DoubleArray(points)
        val centerX = randomCoord(random, GISTestConfig.WORLD_MIN_X, GISTestConfig.WORLD_MAX_X)
        val centerY = randomCoord(random, GISTestConfig.WORLD_MIN_Y, GISTestConfig.WORLD_MAX_Y)
        val radius = 20 + random.nextDouble() * 80

        for (i in 0 until points) {
            val angle = 2 * Math.PI * i / points + random.nextDouble() * 0.3
            xPoints[i] = centerX + Math.cos(angle) * radius
            yPoints[i] = centerY + Math.sin(angle) * radius
        }

        val fillColor = randomColor(random, 80, 180, 0.5)
        val strokeColor = randomColor(random, 60, 160, 1.0)
        val strokeWidth = 1.0 + random.nextDouble() * 2.5
        return GISPolygon(xPoints, yPoints, fillColor, strokeColor, strokeWidth)
    }

    private fun randomCoord(random: Random, min: Double, max: Double): Double {
        return min + (max - min) * random.nextDouble()
    }

    private fun randomColor(random: Random, min: Int, max: Int, alpha: Double): Color {
        val r = min + random.nextInt(max - min + 1)
        val g = min + random.nextInt(max - min + 1)
        val b = min + random.nextInt(max - min + 1)
        return Color.rgb(r, g, b, alpha)
    }
}

